language-plutus-core/plutus-ir/Language/PlutusIR/Purity.hs

```diff
@@ -0,0 +1,88 @@
+saturatesScheme [] TypeSchemeResult{} = True
+saturatesScheme (TermArg _ : args) (TypeSchemeArrow _ sch) = saturatesScheme args sch
+saturatesScheme (TypeArg _ : args) (TypeSchemeAllType _ k) = saturatesScheme args (k Proxy)
+saturatesScheme _ _ = False
+
+-- | Is the given 'BuiltinApp' saturated?
+isSaturated
+    :: forall tyname name uni a term
+    . (HasConstantIn uni term, PLC.GShow uni, PLC.GEq uni, PLC.DefaultUni PLC.<: uni)
+    => DynamicBuiltinNameMeanings term
+    -> BuiltinApp tyname name uni a
+    -> Bool
+isSaturated (DynamicBuiltinNameMeanings means) (BuiltinApp b args) = case b of
+    PLC.BuiltinName _ bn -> withTypedBuiltinName @uni @term bn $ \(TypedBuiltinName _ sch) -> saturatesScheme args sch
+    PLC.DynBuiltinName _ bn -> case Map.lookup bn means of
+        Just (DynamicBuiltinNameMeaning sch _ _) -> saturatesScheme args sch
+        Nothing -> False
+
+-- | View a 'Term' as a 'BuiltinApp' if possible.
+asBuiltinApp :: Term tyname name uni a -> Maybe (BuiltinApp tyname name uni a)
+asBuiltinApp = go []
+    where
+        go argsSoFar = \case
+            Apply _ t arg -> go (TermArg arg:argsSoFar) t
+            TyInst _ t arg -> go (TypeArg arg:argsSoFar) t
+            -- We accumulated the arguments last-first, so reverse them here.
+            Builtin _ b -> Just $ BuiltinApp b (reverse argsSoFar)
+            _ -> Nothing
+
+{- Note [Purity, strictness, and variables]
+Variables in PLC won't have effects: they can have something else substituted for them, but those will be fully evaluated already.
+However, in PIR we have non-strict variable bindings. References to non-strict variables *can* have effects - after all,
+they compile into an application.
+
+So we need to take this into account in our purity calculation. We require the user to tell us which variables are strict, this
+must be *conservative* (i.e. if you don't know, it's non-strict).
+-}
+
+-- | Will evaluating this term have side effects (looping or error)?. This is slightly wider than the definition of a value, as
+-- it includes things that can't be returned from the machine (as they'd be ill-scoped).
+isPure
+    :: (HasConstantIn uni term, PLC.GShow uni, PLC.GEq uni, PLC.DefaultUni PLC.<: uni)
+    => DynamicBuiltinNameMeanings term -> (name -> Strictness) -> Term tyname name uni a -> Bool
+isPure means varStrictness = go
+    where
+        go = \case
+            -- See Note [Purity, strictness, and variables]
+            Var _ n -> case varStrictness n of
+                Strict -> True
+                NonStrict -> False
+            -- These are syntactically values that won't reduce further
+            LamAbs {} -> True
+            TyAbs {} -> True
+            Constant {} -> True
+            IWrap _ _ _ t -> go t
+
+            -- "Stuck" builtin applications won't compute, so won't have effects
+            x | Just bapp@(BuiltinApp _ args) <- asBuiltinApp x ->
+                not $ isSaturated means bapp
```

Comment:
You know, neither have I, and I'm actually not sure that's correct!

Reply:
`$` has lowest precedence possible (`0`), you don't need to learn its relation to other operators, because

- I guess, all (or nearly all) of them have precedence higher than `0`
- even if some operator `@` also has precedence `0`, then `f $ x @ y` will either fail to parse if `@` is not right associative or will parse as you expect `f (x @ y)`

So see `f $ <expr>` -- read `f (<expr>)`.

The only "operator" that I can come up with that behaves differently is `::` (type annotation). I.e. `f $ x :: A` parses as `f x :: A`.